Sources/MongoSwift/MongoClient.swift

```diff
@@ -168,8 +170,23 @@ public class MongoClient {
     /// Default maximum size for connection pools created by this client.
     internal static let defaultMaxConnectionPoolSize = 100
 
-    /// Indicates whether this client has been closed.
-    internal private(set) var isClosed = false
+    /// Indicates the state of this `MongoClient`.
+    internal enum State {
+        /// The client is running and has not been shut down.
+        case upAndRunning
+        /// The client is in the process of shutting down.
+        case shuttingDown
+        /// The client has finished shutting down.
+        case shutDown
+    }
+
+    /// Indicates the state of the client.
+    internal private(set) var state: State
```

Comment:
this + the lock is to provide a little more fine grained and accurate info about what's going in the client at any given point and to let us prevent more operations from getting queued up once shutdown has been called.